### DISCUSSION TITLE
AsyncEngine + asyncpg does not use `executemany` when table.insert().values() is a list of dict

### DISCUSSION THREAD

**Discussion by dnp1** [sqlalchemy/sqlalchemy#7651] [Usage Questions]
### Describe the use case

When executing table.insert().values(list_of_dict)  or table.insert(list_of_dict) sqlalchemy+asyncpg does not use executemany. 

However, passing a list of dict as params to execute make asyncpg use `executemany` properly, why values cannot have the same behaviour?

### Databases / Backends / Drivers targeted

- OS: Linux
- Python: 3.9.9
- SQLAlchemy: 1.4.31
- Database: postgresql 13
- DBAPI : asyncpg

### Example Use

async def example(engine: AsyncEngine):
    async with engine.begin() as conn:
      await conn.execute(table.insert().values(list_of_dict))
      # slow, not using execute many, each dict key is a parameter, why not use 
     
      await conn.execute(table.insert(), list_of_dict)
      # fast, uses execute many

**Comment by zzzeek** [comment]
SQLAlchemy execute() methods lead down into DBAPI cursor.execute() or cursor.executemany() solely based on the format of the second argument, which is the parameter collection.  If the parameter collection is non-present, or is a single dictionary, it uses cursor.execute().  If the parameter collection is a list of dictionaries, it uses cursor.executemany().   This behavior is completely consistent regardless of the kind of SQL construct that's used.

as it turns out, if you do some performance testing you'll find that insert.values() where you pass the list of parameters to the insert() directly is in fact way faster, especially on PostgreSQL dialects, compared to using "executemany".   In fact the SQLAlchemy psycopg2 dialect already has an internal behavior that converts an "executemany" for an insert into a single insert.values(), using the [psycopg2 fast execution helpers](https://www.psycopg.org/docs/extras.html#fast-execution-helpers) - and we will be generalizing this behavior to all the other PostgreSQL dialects including asyncpg in any case, for version 2.0 - the single statement execution with lots of parameter sets is like an order of magnitude faster than the individual statement executions (see the performance stats at https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#psycopg2-dialect-features-execute-values-with-returning-for-insert-statements-by-default).

so at the end of the day in 2.0, both styles of insert() will end up *not* using executemany; it's actually slower.

**Reply by dnp1** [reply]
My practical experience using asyncpg is not the described here.

I can provide you a proper dataset and snippet.

Maybe I'm wrong about the usage or not of executemany, however, there is a performance gap between  conn.execute(query, params) and conn.execute(insert.values())

**Reply by CaselIT** [reply]
> I can provide you a proper dataset and snippet.

examples are always welcome

**Reply by dnp1** [reply]
On my machine, using execute with params takes 40ms while using sqlalchemy values take 160ms.

Param binding is completely different between them, using execute param is possible insert up to 32K rows, since each row is a tuple param, while using "values" it is limited to 32k column values. So 10.000 example not runs with sqlalchemy values.

I'm considering revert to aiopg.


https://gist.github.com/dnp1/ffda22461a2cb5726db284815eae8be9

**Reply by zzzeek** [reply]
> 
> I'm considering revert to aiopg.

if you feel it's constructive to point this out, I could only guess this is meant as some indication that we haven't given you adequate support.    Certainly if you need an immediate change in behavior before we've had time to review your benching results, feel free to revert, although as mentioned in my first answer, SQLAlchemy's asyncpg dialect **does** use executemany() if you simply pass the insert() and the parameters separately, and additionally, "reverting" would not change the behavior you take issue with because I highly doubt aiopg is splitting open insert() objects and converting them to executemany().

I'd like to get as much information as possible about real-world performance scenarios and we most certainly take them all into account, so thanks for providing one.   Any decisions you have to make regarding not using our project our fine but no need to use that here as some kind of indirect communication, thanks.

**Reply by zzzeek** [reply]
I've replied to your result.  the insert.values() approach requires chunking, which SQLAlchemy's "executemany()" adapter, if and when we implement for asyncpg, does transparently.  So indeed asyncpg's executemany seems to be faster than values() when RETURNING is not used, so we will keep that in place.

The insert.values() optimization is most important when we want to use RETURNING, as the ORM usually does, and we can't use a real executemany() approach there.   So we will make the optimization conditional in asyncpg if RETURNING is requested or not.    thanks for the info!

**Reply by dnp1** [reply]
@zzzeek  returing actually does not works using conn.execute(query, params).

Is this expected?

**Reply by zzzeek** [reply]
> @zzzeek returing actually does not works using conn.execute(query, params).
> 
> Is this expected?

RETURNING works with conn.execute().  whether you are hitting a bug in some specific calling form or not using the API correctly is not yet known.

**Reply by dnp1** [reply]
I don't know how  aiopg.sa is implemented, but I could use everything (including large inserts with returning) without having to chunk .

**Reply by zzzeek** [reply]
you used this form:

    conn.execute(insert.values())

directly and it was as fast as execute(insert, values) ?     I dont see how that would be possible

**Reply by zzzeek** [reply]
unless they crack open the insert.values() and convert it.  but they can't use returning with [executemany](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.prepared_stmt.PreparedStatement.executemany) - results are discarded.

**Reply by dnp1** [reply]
@zzzeek  I think it is a bug. I've updated  my snippet
 
 ```
 # works
    async def test_insert_with_values_returing(self):
        ins = table_test_1.insert().values(_1000_).returning(table_test_1.c.id)
        async with self.engine.begin() as conn:
            res = await conn.execute(ins)
            data = res.fetchall()
            self.assertEqual(1000, len(data))

# does not works
    async def test_insert_with_values_as_execute_param(self):
        ins = table_test_1.insert().returning(table_test_1.c.id)
        async with self.engine.begin() as conn:
            res = await conn.execute(ins, _1000_)
            data = res.fetchall()
            self.assertEqual(1000, len(data))
 ```

**Reply by dnp1** [reply]
I'll take a look at aiopg code
